server/src/test/java/org/kaazing/gateway/server/topic/AbstractClusterTopicTest.java

```diff
@@ -0,0 +1,236 @@
+                topicMember1.removeMessageListener(name);
+                listenerCalled.countDown();
+            });
+            t.start();
+            try {
+                t.join();
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        };
+        topicMember1.addMessageListener(m2);
+        topicMember2.publish("msg1");
+        listenerCalled.await();
+        assertFalse(topicMember1.removeMessageListener(name)); // already removed
+        topicMember1.destroy();
+        topicMember2.destroy();
+    }
+
+    @Test
+    public void shouldPubSubFromMessageListeners() throws InterruptedException {
+        CountDownLatch listenersCalled = new CountDownLatch(1);
+        ITopic<String> topic1 = getTopicMember1ForShouldPubSubFromMessageListeners();
+        ITopic<String> topic2 = getTopicMember2ForShouldPubSubFromMessageListeners();
+        MessageListener<String> m1 = message -> topic2.publish("Message sent to topic2");
+        MessageListener<String> m2 = message -> listenersCalled.countDown();
+        topic1.addMessageListener(m1);
+        topic2.addMessageListener(m2);
+
+        topic1.publish("trigger m2");
+        listenersCalled.await();
+        assertEquals(1, topic1.getLocalTopicStats().getPublishOperationCount());
+        assertEquals(1, topic1.getLocalTopicStats().getReceiveOperationCount());
+
+        assertEquals(1, topic2.getLocalTopicStats().getPublishOperationCount());
+        assertEquals(1, topic2.getLocalTopicStats().getReceiveOperationCount());
+        topic1.destroy();
+        topic2.destroy();
+    }
+
+    @Test
+    public void shouldNotDeadlockNestedPublishOnDifferentThread() throws InterruptedException {
+        ITopic<String> topic = getTopicMember1ForShouldNotDeadlockNestedPublishOnDifferentThread();
+        CountDownLatch listenerCalled = new CountDownLatch(10);
+        topic.addMessageListener(message -> {
+            listenerCalled.countDown();
+            Thread t = new Thread(() -> {
+                if (listenerCalled.getCount() > 0) {
+                    topic.publish("Resend: " + message.getMessageObject());
+                }
+            });
+            t.start();
+            try {
+                t.join();
+            } catch (InterruptedException e) {
+                e.printStackTrace();
+            }
+        });
+        topic.publish("KickOff");
+        listenerCalled.await();
+        assertEquals(10, topic.getLocalTopicStats().getPublishOperationCount());
+        assertEquals(10, topic.getLocalTopicStats().getReceiveOperationCount());
+        topic.destroy();
+    }
+
+    @Test(expected = NullPointerException.class)
+    public void shouldNotAddNullMessageListener() {
+        ITopic<String> topic = getTopicMember1ForShouldNotAddNullMessageListener();
+        topic.addMessageListener(null);
+    }
+
+    @Test
+    // TODO take into account the following as limitation or study how the type can be enforced
```

Comment:
Very clever of you to think of this test case! What did you have in mind  for the TODO item? Do you think we really can or should try to do something about it? I think no, since your DefaultClusterTopicTest shows Hazelcast as the same behavior. So I would recommend just removing this TODO comment.